File: app/src/main/java/com/android/mathias/velocity/FragmentRoutes.java
package com.android.mathias.velocity;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import static android.app.Activity.RESULT_OK;

public class FragmentRoutes extends android.support.v4.app.Fragment implements IClickInterface {

    private RecyclerAdapterRoutes mAdapter;
    private List<Route> mListRoutes = new ArrayList<>();
    private List<Route> mListRoutesBackup = null;
    private FloatingActionButton fabCreate = null;

    private Route mTempRoute = null;
    private Snackbar mSnackbar = null;
    private boolean mMoveMode = false;

    private Toolbar mToolbar = null;
    private BottomNavigationView mNavBar = null;

    private static final int REQUEST_ROUTE_DATA = 200;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View routesView = inflater.inflate(R.layout.fragment_routes, container, false);
        initRecyclerView(routesView);
        setHasOptionsMenu(true);
        mToolbar = Objects.requireNonNull(getActivity()).findViewById(R.id.toolbar);
        mNavBar = Objects.requireNonNull(getActivity()).findViewById(R.id.navigation);
        fabCreate = routesView.findViewById(R.id.fab_create_route);
        fabCreate.setOnClickListener(v -> FragmentRoutes.this.handleFabEvent());
        List<Route> routes = DBManager.getRoutes(getContext(), null);
        for (Route r : routes) {
            addRouteCard(r);
        }
        return routesView;
    }

    private void handleFabEvent() {
        startActivityForResult(new Intent(getActivity(), ActivityCreateRoute.class), REQUEST_ROUTE_DATA);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_ROUTE_DATA) {
            if (resultCode == RESULT_OK) {
                Route newRoute = new Route();
                Bundle result = data.getBundleExtra(ActivityCreateRoute.RESULT_BUNDLE);
                double[] start = result.getDoubleArray(ActivityCreateRoute.START_LOC);
                double[] end = result.getDoubleArray(ActivityCreateRoute.END_LOC);
                if (start != null && end != null) {
                    newRoute.setName(result.getString(ActivityCreateRoute.ROUTE_NAME));
                    newRoute.setStartLoc(new LatLng(start[0], start[1]));
                    newRoute.setEndLoc(new LatLng(end[0], end[1]));
                    newRoute.setStartName(result.getString(ActivityCreateRoute.START_LOC_NAME));
                    newRoute.setEndName(result.getString(ActivityCreateRoute.END_LOC_NAME));
                    newRoute.setPos(mListRoutes.size() + 1);
                    DBManager.saveRoute(getContext(), newRoute);
                    addRouteCard(newRoute);
                }
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.clear();
        inflater.inflate(R.menu.menu_routes, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                startActivity(new Intent(getActivity(), ActivitySettings.class));
                break;
            case R.id.action_delete_routes:
                DBManager.deleteAllRoutes(getActivity());
                mListRoutes.clear();
                mAdapter.notifyDataSetChanged();
                break;
            case R.id.action_rearrange:
                mListRoutesBackup = new ArrayList<>(mListRoutes);
                toggleRearrangeMode();
                break;
            case R.id.action_apply:
                persistNewOrder();
                toggleRearrangeMode();
                break;
            case R.id.action_cancel:
                restoreOrderBackup();
                toggleRearrangeMode();
                break;
            case R.id.action_about:
                createRoutesDemoData();
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void toggleRearrangeMode() {
        mMoveMode = !mMoveMode;
        mToolbar.getMenu().clear();
        int menu = mMoveMode ? R.menu.menu_apply_changes : R.menu.menu_routes;
        int color = getResources().getColor(mMoveMode ? R.color.colorAccent : R.color.colorPrimary, Objects.requireNonNull(getActivity()).getTheme());
        int title = mMoveMode ? R.string.title_rearrange : R.string.nav_item_routes;
        mToolbar.inflateMenu(menu);
        mToolbar.setBackgroundColor(color);
        mToolbar.setTitle(title);
        fabCreate.setVisibility(mMoveMode ? View.GONE : View.VISIBLE);
        mNavBar.setVisibility(mMoveMode ? View.GONE : View.VISIBLE);
        mAdapter.setRearrangeMode(mMoveMode);
    }

    private void persistNewOrder() {
        for (int i = 0; i < mListRoutes.size(); i++) {
            DBManager.setRoutePos(getContext(), mListRoutes.get(i).getId(), i);
        }
    }

    private void restoreOrderBackup() {
        mListRoutes.clear();
        mListRoutes.addAll(mListRoutesBackup);
        mAdapter.notifyDataSetChanged();
        mListRoutesBackup.clear();
        mListRoutesBackup = null;
    }

    @Override
    public void onDetach() {
        if (mSnackbar != null) {
            mSnackbar.dismiss();
            mSnackbar = null;
        }
        super.onDetach();
    }

    @Override
    public void itemLongClick(View v, int position) {
        if (!mMoveMode)
            showBottomSheet();
    }

    public void showBottomSheet() {
        BottomSheetModal bottomSheetDialogFragment = new BottomSheetModal();
        bottomSheetDialogFragment.show(Objects.requireNonNull(getActivity())
                .getSupportFragmentManager(), "Bottom Sheet Dialog Fragment");
    }

    private void addRouteCard(Route route) {
        mListRoutes.add(route);
        mAdapter.notifyItemInserted(mListRoutes.size() - 1);
    }

    private void initRecyclerView(View routesView) {
        final RecyclerView rvRoutes = routesView.findViewById(R.id.list_routes);
        mListRoutes.clear();
        mAdapter = new RecyclerAdapterRoutes(mListRoutes, rvRoutes, this);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(Objects.requireNonNull(getActivity()).getApplicationContext());
        rvRoutes.setLayoutManager(layoutManager);
        rvRoutes.setAdapter(mAdapter);
        final Snackbar.Callback sbCallback = new Snackbar.Callback() {
            @Override
            public void onDismissed(Snackbar sb, int event) {
                if (mSnackbar != null && mTempRoute != null) {
                    DBManager.deleteRoute(getContext(), mTempRoute.getId());
                    mSnackbar.removeCallback(this);
                    mSnackbar = null;
                    mTempRoute = null;
                }
                super.onDismissed(sb, event);
            }
        };
        ItemTouchHelper.SimpleCallback ithCallback = new ItemTouchHelper.SimpleCallback(0, 0) {
            @Override
            public void onMoved(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, int fromPos, RecyclerView.ViewHolder target, int toPos, int x, int y) {
                Collections.swap(mListRoutes, fromPos, toPos);
                super.onMoved(recyclerView, viewHolder, fromPos, target, toPos, x, y);
            }

            @Override
            public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
                mAdapter.notifyItemMoved(viewHolder.getAdapterPosition(), target.getAdapterPosition());
                return true;
            }

            @Override
            public int getDragDirs(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
                if (mMoveMode)
                    return ItemTouchHelper.UP | ItemTouchHelper.DOWN;
                else
                    return super.getDragDirs(recyclerView, viewHolder);
            }

            @Override
            public int getSwipeDirs(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
                if (!mMoveMode)
                    return ItemTouchHelper.START | ItemTouchHelper.END;
                else
                    return super.getSwipeDirs(recyclerView, viewHolder);
            }

            @Override
            public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {
                if (mSnackbar != null && mTempRoute != null) {
                    DBManager.deleteRoute(getContext(), mTempRoute.getId());
                    mSnackbar.removeCallback(sbCallback);
                    mSnackbar = null;
                    mTempRoute = null;
                }
                final int idx = viewHolder.getAdapterPosition();
                mTempRoute = mListRoutes.get(idx);
                mListRoutes.remove(idx);
                mAdapter.notifyItemRemoved(idx);
                mSnackbar = Snackbar.make(rvRoutes, "\"" + mTempRoute.getName() + "\" deleted.", Snackbar.LENGTH_LONG).setAction("UNDO", view -> {
                    if (mTempRoute != null) {
                        mSnackbar.removeCallback(sbCallback);
                        mListRoutes.add(idx, mTempRoute);
                        mAdapter.notifyItemInserted(idx);
                        Snackbar.make(rvRoutes, "Restored.", Snackbar.LENGTH_SHORT).show();
                        mSnackbar = null;
                        mTempRoute = null;
                    } else {
                        Snackbar.make(rvRoutes, "Error restoring...", Snackbar.LENGTH_SHORT).show();
                    }
                }).addCallback(sbCallback);
                mSnackbar.show();
            }
        };
        ItemTouchHelper ith = new ItemTouchHelper(ithCallback);
        ith.attachToRecyclerView(rvRoutes);
    }

    private void createRoutesDemoData() {
        for (int i = 1; i <= 5; i++) {
            LatLng startLoc = new LatLng((i + 0.002) * Math.PI, (i + 0.001) * Math.PI);
            LatLng endLoc = new LatLng(i * Math.PI, i * Math.PI);
            Route route = new Route("To location " + i, startLoc, endLoc, "Somewhere " + i, "Nowhere " + i);
            DBManager.saveRoute(getContext(), route);
            addRouteCard(route);
        }
    }
}

interface IClickInterface {
    void itemLongClick(View v, int position);
}

File: app/src/main/java/com/android/mathias/velocity/FragmentRoutes.kt
package com.android.mathias.velocity

import android.app.Activity.RESULT_OK
import android.content.Intent
import android.os.Bundle
import android.view.*
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.android.mathias.velocity.db.DBManager
import com.android.mathias.velocity.ext.IRvClickListener
import com.android.mathias.velocity.model.Route
import com.android.mathias.velocity.util.BottomSheetRoutes
import com.android.mathias.velocity.util.RecyclerAdapterRoutes
import com.google.android.gms.maps.model.LatLng
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import java.util.*

class FragmentRoutes : Fragment(), IRvClickListener {

    private lateinit var mRvRoutes : RecyclerView
    private lateinit var fabCreate: FloatingActionButton
    private lateinit var mAdapter: RecyclerAdapterRoutes
    private lateinit var mToolbar: Toolbar
    private lateinit var mNavBar: BottomNavigationView

    private val mListRoutes = ArrayList<Route>()
    private var mListRoutesBackup: MutableList<Route>? = null
    private var mTempRoute: Route? = null

    private var mSnackbar: Snackbar? = null
    private var mMoveMode = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val routesView = inflater.inflate(R.layout.fragment_routes, container, false)
        setHasOptionsMenu(true)
        initRecyclerView(routesView)
        mToolbar = activity!!.findViewById(R.id.toolbar)
        mNavBar = activity!!.findViewById(R.id.navigation)
        fabCreate = routesView.findViewById(R.id.fab_create_route)
        fabCreate.setOnClickListener { this@FragmentRoutes.handleFabEvent() }
        val routes = DBManager.getRoutes(context!!, null)
        for (r in routes) {
            addRouteCard(r)
        }
        return routesView
    }

    private fun initRecyclerView(routesView: View) {
        mRvRoutes = routesView.findViewById(R.id.list_routes)
        mListRoutes.clear()
        mAdapter = RecyclerAdapterRoutes(mListRoutes, mRvRoutes, this)
        val layoutManager = LinearLayoutManager(activity!!.applicationContext)
        mRvRoutes.layoutManager = layoutManager
        mRvRoutes.adapter = mAdapter
        itemTouchHelper.attachToRecyclerView(mRvRoutes)
    }

    private fun handleFabEvent() {
        startActivityForResult(Intent(activity, ActivityCreateRoute::class.java), REQUEST_ROUTE_DATA)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_ROUTE_DATA) {
            if (resultCode == RESULT_OK) {
                val newRoute = Route()
                val result = data!!.getBundleExtra(ActivityCreateRoute.RESULT_BUNDLE)
                val start = result.getDoubleArray(ActivityCreateRoute.START_LOC)
                val end = result.getDoubleArray(ActivityCreateRoute.END_LOC)
                if (start != null && end != null) {
                    newRoute.name = result.getString(ActivityCreateRoute.ROUTE_NAME)
                    newRoute.startLoc = LatLng(start[0], start[1])
                    newRoute.endLoc = LatLng(end[0], end[1])
                    newRoute.startName = result.getString(ActivityCreateRoute.START_LOC_NAME)
                    newRoute.endName = result.getString(ActivityCreateRoute.END_LOC_NAME)
                    newRoute.pos = mListRoutes.size + 1
                    newRoute.id = DBManager.saveRoute(context!!, newRoute)
                    newRoute.pos = mListRoutes.size
                    addRouteCard(newRoute)
                }
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        menu!!.clear()
        inflater!!.inflate(R.menu.menu_routes, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.action_settings -> startActivity(Intent(activity, ActivitySettings::class.java))
            R.id.action_delete_routes -> {
                DBManager.deleteAllRoutes(context!!)
                mListRoutes.clear()
                mAdapter.notifyDataSetChanged()
            }
            R.id.action_rearrange -> {
                mListRoutesBackup = ArrayList(mListRoutes)
                toggleRearrangeMode()
            }
            R.id.action_apply -> {
                persistNewOrder()
                toggleRearrangeMode()
            }
            R.id.action_cancel -> {
                restoreOrderBackup()
                toggleRearrangeMode()
            }
            R.id.action_about -> createRoutesDemoData()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun toggleRearrangeMode() {
        mMoveMode = !mMoveMode
        mToolbar.menu.clear()
        val menu = if (mMoveMode) R.menu.menu_apply_changes else R.menu.menu_routes
        val color = resources.getColor(if (mMoveMode) R.color.colorAccent else R.color.colorPrimary, activity!!.theme)
        val title = if (mMoveMode) R.string.title_rearrange else R.string.nav_item_routes
        mToolbar.inflateMenu(menu)
        mToolbar.setBackgroundColor(color)
        mToolbar.setTitle(title)
        if (mMoveMode) fabCreate.hide() else fabCreate.show()
        mNavBar.visibility = if (mMoveMode) View.GONE else View.VISIBLE
        mAdapter.setRearrangeMode(mMoveMode)
    }

    private fun persistNewOrder() {
        for (i in mListRoutes.indices) {
            DBManager.setRoutePos(context!!, mListRoutes[i].id, i)
        }
    }

    private fun restoreOrderBackup() {
        mListRoutes.clear()
        mListRoutes.addAll(mListRoutesBackup!!)
        mAdapter.notifyDataSetChanged()
        mListRoutesBackup!!.clear()
        mListRoutesBackup = null
    }

    override fun onDetach() {
        if (mSnackbar != null) {
            mSnackbar!!.dismiss()
            mSnackbar = null
        }
        super.onDetach()
    }

    override fun itemLongClick(v: View, position: Int) {
        if (!mMoveMode) {
            showBottomSheet(mListRoutes[position])
        }
    }

    private fun showBottomSheet(route: Route) {
        val sheet = BottomSheetRoutes().newInstance(route)
        sheet.show(activity!!.supportFragmentManager, "Bottom Sheet Dialog Fragment")
    }

    private fun addRouteCard(route: Route) {
        mListRoutes.add(route)
        mAdapter.notifyItemInserted(mListRoutes.lastIndex)
    }

    private fun createRoutesDemoData() {
        for (i in 1..5) {
            val startLoc = LatLng((i + 0.002) * Math.PI, (i + 0.001) * Math.PI)
            val endLoc = LatLng(i * Math.PI, i * Math.PI)
            val route = Route("To location $i", startLoc, endLoc, "Somewhere $i", "Nowhere $i")
            route.pos = mListRoutes.size
            route.id = DBManager.saveRoute(context!!, route)
            addRouteCard(route)
        }
    }

    private val sbCallback = object : Snackbar.Callback() {
        override fun onDismissed(sb: Snackbar?, event: Int) {
            if (mSnackbar != null && mTempRoute != null) {
                DBManager.deleteRoute(context!!, mTempRoute!!.id)
                mSnackbar!!.removeCallback(this)
                for (i: Int in mTempRoute!!.pos until mListRoutes.size) {
                    mListRoutes[i].pos--
                    DBManager.setRoutePos(context!!, mListRoutes[i].id, mListRoutes[i].pos)
                }
                mSnackbar = null
                mTempRoute = null
            }
            super.onDismissed(sb, event)
        }
    }

    private val itemTouchHelper = ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, 0) {
        override fun onMoved(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder,
                             fromPos: Int, target: RecyclerView.ViewHolder, toPos: Int, x: Int, y: Int) {
            Collections.swap(mListRoutes, fromPos, toPos)
            super.onMoved(recyclerView, viewHolder, fromPos, target, toPos, x, y)
        }

        override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
            mAdapter.notifyItemMoved(viewHolder.adapterPosition, target.adapterPosition)
            return true
        }

        override fun getDragDirs(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
            return if (mMoveMode)
                ItemTouchHelper.UP or ItemTouchHelper.DOWN
            else
                super.getDragDirs(recyclerView, viewHolder)
        }

        override fun getSwipeDirs(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
            return if (!mMoveMode)
                ItemTouchHelper.START or ItemTouchHelper.END
            else
                super.getSwipeDirs(recyclerView, viewHolder)
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
            if (mSnackbar != null && mTempRoute != null) {
                DBManager.deleteRoute(context!!, mTempRoute!!.id)
                mSnackbar!!.removeCallback(sbCallback)
                for (i: Int in mTempRoute!!.pos until mListRoutes.size) {
                    mListRoutes[i].pos--
                    DBManager.setRoutePos(context!!, mListRoutes[i].id, mListRoutes[i].pos)
                }
                mSnackbar = null
                mTempRoute = null
            }
            val idx = viewHolder.adapterPosition
            mTempRoute = mListRoutes[idx]
            mListRoutes.removeAt(idx)
            mAdapter.notifyItemRemoved(idx)
            mSnackbar = Snackbar.make(mRvRoutes, "\"${mTempRoute!!.name}\" deleted.", Snackbar.LENGTH_LONG).setAction("UNDO") {
                if (mTempRoute != null) {
                    mSnackbar!!.removeCallback(sbCallback)
                    mListRoutes.add(idx, mTempRoute!!)
                    mAdapter.notifyItemInserted(idx)
                    Snackbar.make(mRvRoutes, "Restored.", Snackbar.LENGTH_SHORT).show()
                    mSnackbar = null
                    mTempRoute = null
                } else {
                    Snackbar.make(mRvRoutes, "Error restoring...", Snackbar.LENGTH_SHORT).show()
                }
            }.addCallback(sbCallback)
            mSnackbar!!.show()
        }
    })

    companion object {
        private const val REQUEST_ROUTE_DATA = 200
    }
}

